src/memray/_memray/macho_shenanigans.cpp

```diff
@@ -0,0 +1,145 @@
+        const section_t* section,
+        uintptr_t slide,
+        const DynamicInfoTable& dyninfo_table,
+        bool restore_original)
+{
+    auto symbol_addr_table = reinterpret_cast<void**>(slide + section->addr);
+    for (unsigned int i = 0; i < section->size / sizeof(void*); i++) {
+        const char* symbol_name = dyninfo_table.getSymbol(section->reserved1, i);
+        if (!symbol_name || !symbol_name[0] || !symbol_name[1]) {
+            continue;
+        }
+#define FOR_EACH_HOOKED_FUNCTION(hookname)                                                              \
+    if (strcmp(hooks::hookname.d_symbol, symbol_name + 1) == 0) {                                       \
+        patch_symbol(                                                                                   \
+                hooks::hookname,                                                                        \
+                &intercept::hookname,                                                                   \
+                symbol_name,                                                                            \
+                symbol_addr_table + i,                                                                  \
+                section->size,                                                                          \
+                restore_original);                                                                      \
+        continue;                                                                                       \
+    }
+        MEMRAY_HOOKED_FUNCTIONS
+#undef FOR_EACH_HOOKED_FUNCTION
+    }
+}
+
+static void
+patch_symbols_in_shared_object(const struct mach_header* header, intptr_t slide, bool restore_original)
+{
+    Dl_info info;
+    if (dladdr(header, &info) == 0) {
+        return;
+    }
+
+    const char* image_name = info.dli_fname;
+    if (strstr(image_name, "memray.cpython") || strstr(image_name, "/dyld")
+        || strstr(image_name, "dyld.dylib"))
+    {
+        LOG_IF_DEBUG("Skipping patching symbols for " << image_name);
+        return;
+    }
+    LOG_IF_DEBUG("Patching symbols for " << image_name);
+
+    auto segment_filter = [](const char* seg_name) {
+        return strcmp(seg_name, SEG_DATA) == 0 || strcmp(seg_name, SEG_DATA_CONST) == 0
+               || strcmp(seg_name, SEG_AUTH_CONST) == 0;
+    };
+
+    DynamicInfoTable dyninfo_table(header, slide, segment_filter);
+    if (!dyninfo_table) {
+        LOG_IF_DEBUG("Could not construct dynamic information table" << image_name);
+        return;
+    }
+
+    LOG_IF_DEBUG("Found " << dyninfo_table.segments.size() << " data segments");
+
+    for (const auto& segment_cmd : dyninfo_table.segments) {
+        const auto current_seg = reinterpret_cast<uintptr_t>(segment_cmd);
+        auto section_head = reinterpret_cast<const section_t*>(current_seg + sizeof(segment_command_t));
+        LOG_IF_DEBUG("Considering segment " << segment_cmd->segname);
```

Comment:
Any risk of getting a `segname` that's not null-terminated, or anything like that?

Reply:
No, is supposed to be a 16-character max null-terminated string always.